File: exchanges/kis_api.py
import json
import requests
from .base import BaseExchange

class KISClient(BaseExchange):
    def __init__(self, url_base, app_key, app_secret, cano, acnt_prdt_cd):
        self.url_base = url_base
        self.app_key = app_key
        self.app_secret = app_secret
        self.cano = cano
        self.acnt_prdt_cd = acnt_prdt_cd
        self.access_token = None
        self.token_expiry = 0
        self._auth()

    def _auth(self):
        """
        Access Token 발급 및 캐싱
        단독 모바일 구동 시 서버가 며칠씩 켜져 있을 때 24시간 이후 토큰이 만료되어 401 오류가 발생하는 것을 방지하기 위해,
        안전하게 23시간(82800초) 경과 시 토큰을 자동 재발급받도록 만료 시간(token_expiry)을 함께 관리합니다.
        """
        import os, time
        # URL에 vts가 포함되면 모의투자용 토큰 파일 사용
        prefix = "mock_" if "vts" in self.url_base else "real_"
        token_file = f"kis_token_{prefix}.txt"
        
        # 1. 파일에 유효한 토큰이 있으면 재사용 (유효기간 24시간, 넉넉히 23시간으로 계산)
        if os.path.exists(token_file):
            file_mtime = os.path.getmtime(token_file)
            if time.time() - file_mtime < 82800:
                with open(token_file, "r") as f:
                    self.access_token = f.read().strip()
                    self.token_expiry = file_mtime + 82800
                    return

        # 2. 토큰이 없거나 만료되었으면 새로 발급
        headers = {"content-type": "application/json"}
        body = {
            "grant_type": "client_credentials",
            "appkey": self.app_key,
            "appsecret": self.app_secret,
        }
        url = f"{self.url_base}/oauth2/tokenP"
        res = requests.post(url, headers=headers, data=json.dumps(body), timeout=5)
        if res.status_code == 200:
            self.access_token = res.json().get("access_token")
            # 24시간 토큰이므로 23시간(82800초) 뒤를 만료 시점으로 기록하여 사전 갱신 유도
            self.token_expiry = time.time() + 82800
            # 발급받은 토큰을 파일에 저장
            with open(token_file, "w") as f:
                f.write(self.access_token)
        else:
            raise Exception(f"KIS Auth Failed: {res.text}")

    def _get_headers(self, tr_id: str):
        """
        API 요청 시 공통으로 사용되는 헤더를 반환합니다.
        요청 직전에 현재 시간과 token_expiry를 비교하여, 만료 시점이 지났으면 즉각 _auth()를 재호출합니다.
        이를 통해 며칠 동안 봇을 켜두어도 토큰 만료 에러 없이 무중단으로 동작할 수 있습니다.
        """
        import time
        if not self.access_token or time.time() >= self.token_expiry:
            self._auth()
            
        # 모의투자인 경우 TR_ID의 첫 글자 'T'를 'V'로 변경 (예: TTTC8434R -> VTTC8434R)
        if "vts" in self.url_base and tr_id.startswith("T"):
            tr_id = "V" + tr_id[1:]
            
        return {
            "content-type": "application/json",
            "authorization": f"Bearer {self.access_token}",
            "appkey": self.app_key,
            "appsecret": self.app_secret,
            "tr_id": tr_id,
        }

    def get_balance(self):
        """주식 잔고 조회 (TTTC8434R / VTTC8434R)"""
        path = "/uapi/domestic-stock/v1/trading/inquire-balance"
        url = f"{self.url_base}{path}"
        headers = self._get_headers("TTTC8434R")
        params = {
            "CANO": self.cano,
            "ACNT_PRDT_CD": self.acnt_prdt_cd,
            "AFHR_FLPR_YN": "N",
            "OFL_YN": "",
            "INQR_DVSN": "02",
            "UNPR_DVSN": "01",
            "FUND_STTL_ICLD_YN": "N",
            "FNCG_AMT_AUTO_RDPT_YN": "N",
            "PRCS_DVSN": "01",
            "CTX_AREA_FK100": "",
            "CTX_AREA_NK100": "",
        }
        res = requests.get(url, headers=headers, params=params)
        return res.json()

    def get_current_price(self, symbol: str):
        """현재가 조회. FHKST01010100 TR_ID는 실전/모의투자에 공통으로 사용됩니다."""
        path = "/uapi/domestic-stock/v1/quotations/inquire-price"
        url = f"{self.url_base}{path}"
        headers = self._get_headers("FHKST01010100")
        params = {"FID_COND_MRKT_DIV_CODE": "J", "FID_INPUT_ISCD": symbol}
        res = requests.get(url, headers=headers, params=params)
        return res.json()

    def _place_order(self, symbol: str, price: int, quantity: int, is_buy: bool):
        """
        주문 접수 로직.
        초당 요청 횟수 제한(Rate Limit) 등 일시적인 API 서버 거부나 오류 발생 시,
        즉시 실패 처리하지 않고 점진적으로 대기 시간(Incremental Backoff)을 늘려가며 재시도하여 안정성을 극대화합니다.
        """
        import time
        path = "/uapi/domestic-stock/v1/trading/order-cash"
        url = f"{self.url_base}{path}"
        tr_id = "TTTC0802U" if is_buy else "TTTC0801U"
        headers = self._get_headers(tr_id)
        body = {
            "CANO": self.cano,
            "ACNT_PRDT_CD": self.acnt_prdt_cd,
            "PDNO": symbol,
            "ORD_DVSN": "00", # 00: 지정가
            "ORD_QTY": str(quantity),
            "ORD_UNPR": str(price),
        }
        
        max_retries = 5
        base_wait = 1.0
        for attempt in range(max_retries):
            res = requests.post(url, headers=headers, data=json.dumps(body))
            data = res.json()
            
            rt_cd = str(data.get("rt_cd", "0"))
            msg1 = data.get("msg1", "")
            msg_cd = data.get("msg_cd", "")
            
            # KIS Rate Limit에 도달했을 때 점진적 대기(Exponential/Incremental Backoff)를 수행합니다.
            # 서버 과부하를 막고, 일정 시간 뒤 정상적으로 요청이 수락되도록 유도합니다.
            if rt_cd != "0" and ("초과" in msg1 or "건수" in msg1 or "EGW00" in msg_cd):
                wait_time = base_wait + (attempt * 0.8) # 1.0, 1.8, 2.6, 3.4, 4.2초...
                print(f"[Rate Limit] KIS API 제한 도달 ({msg1}). {wait_time:.1f}초 대기 후 재시도... ({attempt+1}/{max_retries})")
                time.sleep(wait_time)
                continue
                
            return data
            
        return res.json()

    def buy_limit(self, symbol: str, price: int, quantity: int):
        return self._place_order(symbol, price, quantity, is_buy=True)

    def sell_limit(self, symbol: str, price: int, quantity: int):
        return self._place_order(symbol, price, quantity, is_buy=False)

    def get_open_orders(self):
        """주식 정정/취소 가능 주문(미체결) 조회 (TTTC8036R / VTTC8036R)"""
        path = "/uapi/domestic-stock/v1/trading/inquire-psbl-rvsecnl"
        url = f"{self.url_base}{path}"
        headers = self._get_headers("TTTC8036R")
        params = {
            "CANO": self.cano,
            "ACNT_PRDT_CD": self.acnt_prdt_cd,
            "CTX_AREA_FK100": "",
            "CTX_AREA_NK100": "",
            "INQR_DVSN_1": "0",  # 0:전체
            "INQR_DVSN_2": "0",  # 0:전체
        }
        res = requests.get(url, headers=headers, params=params)
        return res.json()
        
    def get_daily_orders(self, start_dt: str, end_dt: str):
        """국내주식 일별 주문체결조회 (TTTC8001R / VTTC8001R)"""
        path = "/uapi/domestic-stock/v1/trading/inquire-daily-ccld"
        url = f"{self.url_base}{path}"
        headers = self._get_headers("TTTC8001R")
        params = {
            "CANO": self.cano,
            "ACNT_PRDT_CD": self.acnt_prdt_cd,
            "INQR_STRT_DT": start_dt,
            "INQR_END_DT": end_dt,
            "SLL_BUY_DVSN_CD": "00",  # 00:전체
            "INQR_DVSN": "00",        # 00:역순
            "PDNO": "",
            "CCLD_DVSN": "01",        # 01:체결, 02:미체결
            "ORD_GNO_BRNO": "",
            "ODNO": "",
            "INQR_DVSN_3": "00",
            "INQR_DVSN_1": "",
            "CTX_AREA_FK100": "",
            "CTX_AREA_NK100": "",
        }
        res = requests.get(url, headers=headers, params=params)
        return res.json()

    def cancel_order(self, order_id: str, symbol: str = "042660"):
        """
        주식 주문 취소 (TTTC0803U / VTTC0803U)
        - order_id: 취소할 원주문번호 (ODNO)
        """
        path = "/uapi/domestic-stock/v1/trading/order-rvsecnml"
        url = f"{self.url_base}{path}"
        headers = self._get_headers("TTTC0803U")
        body = {
            "CANO": self.cano,
            "ACNT_PRDT_CD": self.acnt_prdt_cd,
            "KRX_FWDG_ORD_ORGNO": "",
            "ORGN_ODNO": order_id,
            "ORD_DVSN": "00",          # 00: 지정가
            "RVSE_CNCL_DVSN_CD": "02", # 02: 취소 (01: 정정)
            "ORD_QTY": "0",            # 0: 전량 취소
            "ORD_UNPR": "0",
            "QTY_ALL_ORD_YN": "Y",
            "PDNO": symbol
        }
        import json
        res = requests.post(url, headers=headers, data=json.dumps(body))
        print(f"KIS Cancel Response: {res.text}")
        return res.json()

File: mobile_app/lib/api/kis_api.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'package:shared_preferences/shared_preferences.dart';

class KisApi {
  final String appKey;
  final String appSecret;
  final String cano;
  final bool isMock;
  final String baseUrl;

  String? _token;
  DateTime? _tokenExpiry;

  KisApi({
    required this.appKey,
    required this.appSecret,
    required this.cano,
    this.isMock = true,
  }) : baseUrl = isMock 
          ? 'https://openapivts.koreainvestment.com:29443'
          : 'https://openapi.koreainvestment.com:9443';

  /// KIS OAuth2 토큰 발급 및 캐싱
  /// 토큰 만료 1시간 전까지는 캐시된 토큰을 재사용하여 Rate Limit(1분 1회) 방어
  Future<void> auth() async {
    // API 키가 미등록이면 토큰 발급 시도 자체를 하지 않음
    if (appKey.isEmpty || appSecret.isEmpty) return;

    final prefs = await SharedPreferences.getInstance();
    final cacheKey = isMock ? 'kis_token_mock' : 'kis_token_real';
    final expiryKey = isMock ? 'kis_token_expiry_mock' : 'kis_token_expiry_real';

    // SharedPreferences에 캐싱된 토큰이 있으면 메모리로 복원
    if (_token == null) {
      _token = prefs.getString(cacheKey);
      final expStr = prefs.getString(expiryKey);
      if (expStr != null) _tokenExpiry = DateTime.tryParse(expStr);
    }

    // 유효 기간이 1시간 이상 남아있으면 캐시 토큰 사용
    if (_token != null && _tokenExpiry != null) {
      if (DateTime.now().isBefore(_tokenExpiry!.subtract(const Duration(hours: 1)))) {
        return;
      }
    }

    // 신규 토큰 발급
    final url = Uri.parse('$baseUrl/oauth2/tokenP');
    final response = await http.post(
      url,
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode({
        "grant_type": "client_credentials",
        "appkey": appKey,
        "appsecret": appSecret
      }),
    );

    if (response.statusCode == 200) {
      final data = jsonDecode(response.body);
      _token = data['access_token'];
      _tokenExpiry = DateTime.now().add(const Duration(hours: 23));
      
      await prefs.setString(cacheKey, _token!);
      await prefs.setString(expiryKey, _tokenExpiry!.toIso8601String());
    } else {
      throw Exception('토큰 발급 실패: ${response.body}');
    }
  }

  /// API 요청용 공통 헤더 생성 (토큰 자동 갱신 포함)
  Future<Map<String, String>> _buildHeaders(String trId) async {
    await auth();
    if (_token == null) throw Exception('토큰이 발급되지 않았습니다.');
    return {
      "Content-Type": "application/json",
      "authorization": "Bearer $_token",
      "appkey": appKey,
      "appsecret": appSecret,
      "tr_id": trId,
      "custtype": "P",
    };
  }

  /// 잔고 조회
  Future<Map<String, dynamic>> getBalance() async {
    final trId = isMock ? 'VTTC8434R' : 'TTTC8434R';
    final url = Uri.parse('$baseUrl/uapi/domestic-stock/v1/trading/inquire-balance'
        '?CANO=$cano&ACNT_PRDT_CD=01&AFHR_FLPR_YN=N&OFL_YN=N&INQR_DVSN=02&UNPR_DVSN=01&FUND_STTL_ICLD_YN=N&FNCG_AMT_AUTO_RDPT_YN=N&PRCS_DVSN=00&CTX_AREA_FK100=&CTX_AREA_NK100=');
    
    final response = await http.get(url, headers: await _buildHeaders(trId));
    if (response.statusCode == 200) {
      return jsonDecode(response.body);
    } else {
      throw Exception('잔고 조회 실패: ${response.body}');
    }
  }

  /// 지정가 매수/매도 주문
  /// EXG_DVSN_CD: NXT(넥스트레이드 직접 지정)로 프리마켓/정규장 모두 대응
  Future<Map<String, dynamic>> placeOrder({
    required String orderType,
    required String symbol,
    required int price,
    required int qty,
  }) async {
    String trId;
    if (orderType == 'buy') {
      trId = isMock ? 'VTTC0802U' : 'TTTC0802U';
    } else {
      trId = isMock ? 'VTTC0801U' : 'TTTC0801U';
    }

    final url = Uri.parse('$baseUrl/uapi/domestic-stock/v1/trading/order-cash');
    final response = await http.post(
      url,
      headers: await _buildHeaders(trId),
      body: jsonEncode({
        "CANO": cano,
        "ACNT_PRDT_CD": "01",
        "PDNO": symbol,
        "ORD_DVSN": "00", // 00: 지정가
        "ORD_QTY": qty.toString(),
        "ORD_UNPR": price.toString(),
        // 넥스트레이드(NXT) 전용 시장 코드로 강제 지정
        "EXG_DVSN_CD": "NXT"
      }),
    );

    if (response.statusCode == 200) {
      return jsonDecode(response.body);
    } else {
      throw Exception('주문 실패: ${response.body}');
    }
  }

  /// 미체결 주문 내역 조회
  Future<Map<String, dynamic>> getOpenOrders() async {
    final trId = isMock ? 'VTTC8036R' : 'TTTC8036R';
    final url = Uri.parse('$baseUrl/uapi/domestic-stock/v1/trading/inquire-psbl-rvsecncl'
        '?CANO=$cano&ACNT_PRDT_CD=01&CTX_AREA_FK100=&CTX_AREA_NK100=&INQR_DVSN_1=0&INQR_DVSN_2=0');
    final response = await http.get(url, headers: await _buildHeaders(trId));
    if (response.statusCode == 200) {
      return jsonDecode(response.body);
    } else {
      throw Exception('미체결 조회 실패: ${response.body}');
    }
  }

  /// 당일 체결 내역 조회
  Future<Map<String, dynamic>> getExecutionHistory() async {
    final trId = isMock ? 'VTTC8001R' : 'TTTC8001R';
    final now = DateTime.now();
    final todayStr = "${now.year}${now.month.toString().padLeft(2, '0')}${now.day.toString().padLeft(2, '0')}";
    final url = Uri.parse('$baseUrl/uapi/domestic-stock/v1/trading/inquire-daily-ccld'
        '?CANO=$cano&ACNT_PRDT_CD=01&INQR_STRT_DT=$todayStr&INQR_END_DT=$todayStr&SLL_BUY_DVSN_CD=00&INQR_DVSN=00&PDNO=&CCLD_DVSN=01&ORD_GNO_BRNO=&ODNO=&INQR_DVSN_3=00&INQR_DVSN_1=&CTX_AREA_FK100=&CTX_AREA_NK100=');
    final response = await http.get(url, headers: await _buildHeaders(trId));
    if (response.statusCode == 200) {
      return jsonDecode(response.body);
    } else {
      throw Exception('체결 내역 조회 실패: ${response.body}');
    }
  }

  /// 기간별 실현 손익 조회
  Future<Map<String, dynamic>> getRealizedProfit({
    required String startDate,
    required String endDate,
  }) async {
    final trId = isMock ? 'VTTC8495R' : 'TTTC8495R';
    final url = Uri.parse('$baseUrl/uapi/domestic-stock/v1/trading/inquire-period-trade-profit'
        '?CANO=$cano&ACNT_PRDT_CD=01&INQR_STRT_DT=$startDate&INQR_END_DT=$endDate&SLL_BUY_DVSN_CD=00&INQR_DVSN=00&PDNO=&CTX_AREA_FK100=&CTX_AREA_NK100=');
    final response = await http.get(url, headers: await _buildHeaders(trId));
    if (response.statusCode == 200) {
      return jsonDecode(response.body);
    } else {
      throw Exception('실현 손익 조회 실패: ${response.body}');
    }
  }

  /// 주식 주문 취소 (TTTC0803U: 실전 / VTTC0803U: 모의)
  Future<Map<String, dynamic>> cancelOrder(String orderId, {String symbol = "042660"}) async {
    final trId = isMock ? 'VTTC0803U' : 'TTTC0803U';
    final url = Uri.parse('$baseUrl/uapi/domestic-stock/v1/trading/order-rvsecncl');
    final body = jsonEncode({
      "CANO": cano,
      "ACNT_PRDT_CD": "01",
      "KRX_FWDG_ORD_ORGNO": "",
      "ORGN_ODNO": orderId,
      "ORD_DVSN": "00",
      "RVSE_CNCL_DVSN_CD": "02",
      "ORD_QTY": "0",
      "ORD_UNPR": "0",
      "QTY_ALL_ORD_YN": "Y",
      "PDNO": symbol
    });

    final response = await http.post(url, headers: await _buildHeaders(trId), body: body);
    if (response.body.trim().isEmpty) {
      return {"rt_cd": "9", "msg_cd": "EMPTY", "msg1": "빈 응답 (HTTP ${response.statusCode})"};
    }
    try {
      return jsonDecode(response.body);
    } catch (e) {
      return {"rt_cd": "9", "msg_cd": "PARSE_ERROR", "msg1": "파싱 에러: ${response.statusCode}"};
    }
  }

  /// 현재가 조회
  Future<int> getCurrentPrice(String symbol) async {
    final trId = 'FHKST01010100'; // 시세 조회는 실전/모의 동일
    final url = Uri.parse('$baseUrl/uapi/domestic-stock/v1/quotations/inquire-price?FID_COND_MRKT_DIV_CODE=J&FID_INPUT_ISCD=$symbol');
    final response = await http.get(url, headers: await _buildHeaders(trId));
    if (response.statusCode == 200) {
      final data = jsonDecode(response.body);
      return int.tryParse(data['output']?['stck_prpr']?.toString() ?? '0') ?? 0;
    } else {
      throw Exception('현재가 조회 실패: ${response.body}');
    }
  }
}
